Model transactions part 2

alflennik opened the pull request:
The recently merged PR https://github.com/w3c/aria-at-app/pull/922 added support for transactions, and this follow-up PR actually utilizes transactions throughout the app.

howard-e commented on server/resolvers/TestPlanReportOperations/updateTestPlanReportTestPlanVersionResolver.js:
```diff
@@ -4,19 +4,20 @@ const { AuthenticationError } = require('apollo-server-express');
-} = require('../../models/services.deprecated/TestPlanVersionService');
-const { testResults } = require('../TestPlanRun');
+} = require('../../models/services/TestPlanVersionService');
+const { testResultsResolver } = require('../TestPlanRun');
```
This value being exported does not exist

howard-e commented on server/resolvers/TestPlanReportOperations/updateTestPlanReportTestPlanVersionResolver.js:
```diff
@@ -4,19 +4,20 @@ const { AuthenticationError } = require('apollo-server-express');
-const { findOrCreateTestResult } = require('../TestPlanRunOperations');
-const { submitTestResult, saveTestResult } = require('../TestResultOperations');
+} = require('../../models/services/TestPlanRunService');
+const { findOrCreateTestResultResolver } = require('../TestPlanRunOperations');
```
This value being exported does not exist

howard-e commented on server/resolvers/TestPlanVersion/testPlanReportsResolver.js:
```diff
@@ -1,28 +1,19 @@
-            order: [['createdAt', 'desc']]
-        }
-    );
+    const reports = await getTestPlanReports({
```
This change is great to see! No more "counting" parameters!

alflennik replied:
Yes!

howard-e commented on server/resolvers/deleteCollectionJobResolver.js:
```diff
@@ -1,14 +1,15 @@
 const {
-    deleteCollectionJob
-} = require('../models/services.deprecated/CollectionJobService');
+    removeCollectionJobById
```
This sparks a thought on being consistent with API verb usage. So would renaming this resolver and others be helpful in the future to `remove[Entity]Resolver`?

It's outside the purpose of this PR but I'm thinking the consistency may make it easier when talking about certain files/functions as a future thought.

alflennik replied:
Yes I definitely think that would be a win in a future PR!

howard-e replied:
Great, let's follow that up in a separate issue!

howard-e commented on server/resolvers/usersResolver.js:
```diff
@@ -1,9 +1,12 @@
     // No authentication since participation is public!
-    return getUsers(null, {}, null, null, null, null, {
-        order: [['username', 'ASC']]
+    return getUsers({
```
Seems `userAttributes`, `roleAttributes`, `atAttributes` and `testPlanRunAttributes` should be set to `[]` to maintain the current behavior

alflennik replied:
Thanks for noting that, I set them to null which should have exactly the same effect as now.

howard-e commented on server/scripts/import-tests/index.js:
```diff
@@ -114,12 +109,13 @@ const importTestPlanVersions = async () => {
+                { transaction }
             )
-        ).rows[0].nextval;
+        )[0][0].nextval;
```
This might be confusing, can we use some additional defined parameters here?

alflennik replied:
Yeah I was considering doing it that way. Done.

howard-e replied:
Thanks! Appreciate that change being done and easier to follow than what is currently on `main` as well

howard-e commented on server/tests/util/mock-automation-scheduler-server.js:
```diff
@@ -6,7 +6,7 @@ const {
 const { default: axios } = require('axios');
 const { gql } = require('apollo-server-core');
-const { query } = require('../util/graphql-test-utilities');
+const apolloServer = require('../../graphql-server');
```
I'm unsure I follow why make this change. Could you explain?

alflennik replied:
This code runs during local development as part of the dev server (you can see it in server/app.js) so using a method intended exclusively for testing could potentially cause issues. I don't think there are any issues right now but it's more about preventing possible issues.

howard-e replied:
Gotcha!

howard-e reviewed (commented):
@alflennik this is great! It was quite a bit to review so I can only imagine how much time it took just putting this all together. I'm happy we're getting the API shape we've wanted and even more excited about the doors this can open for some better fine-tuned test suites.

I've done my best to be as thorough as I could, given the weight of the change, and so I've left several in-line comments.

Going through this, with all the instances of transaction now being included in a lot of methods is if there's a way a global transaction value could be used instead.

I do also think it would be helpful if someone outside of myself or Paul could take a pass over the automation related changes to make sure everything is as expected there.

Paul-Clue reviewed (approved):
Looking good! I tested the functionality of the entire app and didn't encounter anything breaking. After fixing the merge conflicts and taking care of Howard's suggestions and concerns, this should be ready to merge. Approved.

alflennik commented on server/models/services/TestResultWriteService.js:
```diff
@@ -1,43 +1,58 @@
+            context
+        );
+        const finalizedTestResults = await getFinalizedTestResults(
+            { ...testPlanReport },
```
Nice, definite improvement

alflennik replied:
Done.

alflennik commented on server/middleware/transactionMiddleware.test.js:
```diff
@@ -0,0 +1,82 @@
+
+        expect(beforeSuccessfulUser).toBe(null);
+        expect(beforeFailedUser).toBe(null);
+        expect(successfulUser).not.toBe(null);
```
Nice idea!

alflennik replied:
Done.

alflennik commented on server/migrations/20220929155029-addMetricsValues.js:
```diff
@@ -62,7 +74,11 @@ module.exports = {
+                await updateTestPlanReportById({
+                    id: testPlanReport.id,
+                    values: updateParams,
+                    t: false
```
Wow, good catch. Two catches actually.

alflennik replied:
Done.

alflennik commented on server/services/PopulatedData/populateData.js:
```diff
@@ -24,7 +24,12 @@ const getTests = require('../../models/services.deprecated/TestsService');
-const populateData = async (locationOfData, { preloaded } = {}) => {
+const populateData = async (
+    locationOfData,
+    { transaction, preloaded } = {}
```
It is a required field. It can't be defaulted because that would create a divergent database state which would differ with the rest of the request.

In the case that the transaction is undefined all the model methods will raise an error like this:

```
Please provide a transaction via the "transaction" field or pass a value of false to specify that a transaction is not needed.
```

howard-e replied:
Got it!

alflennik commented on server/resolvers/ScenarioResult/assertionResultsResolver.js:
```diff
@@ -1,6 +1,10 @@
+const assertionResultsResolver = (
+    scenarioResult,
+    { priority },
+    context // eslint-disable-line no-unused-vars
```
Since the context field in GraphQL has become required for the vast majority of the resolvers, I did a pass of the server to make sure that resolvers are always called with the context variable set. I think that makes the API for using a resolver a lot more consistent and clear.

For the few resolvers that don't currently need the context I made this change just to make the expectation clear that context should be passed in now to prevent issues later.

howard-e replied:
> For the few resolvers that don't currently need the context I made this change just to make the expectation clear that context should be passed in now to prevent issues later.

Gotcha. That seems fine with me

alflennik commented on server/middleware/transactionMiddleware.js:
```diff
@@ -0,0 +1,59 @@
+
+const transactions = {};
+
+const middleware = async (req, res, next) => {
```
Sure thing, done.

alflennik commented:
@howard-e this should be ready for another look. [Recent changes are here](https://github.com/w3c/aria-at-app/pull/940/files/909f7dbe3c22841aa778e46419d8dd568a59d402..b8756b27a60eb53ba8a5f728bf74ddb92dfc9eb6#diff-f99189f59b427fd3723c3a41e08ae8af915ba05ef250af5ac2ad7783c6dd928b), I addressed your feedback and resolved the conflicts by bringing in the latest code from main.

alflennik commented on server/migrations/20230608171911-moveTestPlanReportValuesToTestPlanVersion.js:
```diff
@@ -3,25 +3,25 @@
+} = require('../models/services/TestPlanReportService');
 const populateData = require('../services/PopulatedData/populateData');
-const { testResults } = require('../resolvers/TestPlanRun');
+const { testResultsResolver } = require('../resolvers/TestPlanRun');
```
Thank you for finding those, they're all fixed now.

alflennik commented on server/models/CollectionJob.js:
```diff
@@ -9,7 +9,8 @@ module.exports = function (sequelize, DataTypes) {
                 allowNull: false,
-                primaryKey: true
+                primaryKey: true,
+                autoIncrement: true
```
Thanks! Done.

howard-e commented on server/resolvers/TestPlanReport/draftTestPlanRunsResolver.js:
```diff
@@ -1,4 +1,8 @@
-const draftTestPlanRunsResolver = testPlanReport => {
+const draftTestPlanRunsResolver = (
+    testPlanReport,
+    args, // eslint-disable-line no-unused-vars
+    context // eslint-disable-line no-unused-vars
```
Answered by https://github.com/w3c/aria-at-app/pull/940#discussion_r1520480478

howard-e reviewed (approved):
@alflennik looks good. I took another pass and also confirmed my feedback has been addressed. Getting to address this longstanding tech debt of the API structure in addition to setting up for a better testing environment is a huge plus!

I know there are other PRs that depend on or may conflict with this change. Right now, I'm unsure of the sequencing but I figure you may have a better idea. So I'll leave the merging step up to you when ready.
